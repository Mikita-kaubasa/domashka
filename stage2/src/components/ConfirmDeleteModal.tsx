/**
 * @fileoverview Confirmation modal component for user deletion.
 * This component provides a simple confirmation dialog for deleting users
 * with keyboard navigation support and accessibility features.
 * 
 * @module ConfirmDeleteModal
 * @requires React
 */

import React from 'react';

/**
 * Props for the ConfirmDeleteModal component.
 * 
 * @interface ConfirmDeleteModalProps
 * @property {boolean} open - Whether the confirmation modal is currently open
 * @property {() => void} onCancel - Function to call when deletion is cancelled
 * @property {() => void} onConfirm - Function to call when deletion is confirmed
 * @property {string} userName - Name of the user to be deleted (displayed in confirmation message)
 */
interface ConfirmDeleteModalProps {
    open: boolean;
    onCancel: () => void;
    onConfirm: () => void;
    userName: string;
}

/**
 * Confirmation modal for user deletion.
 * 
 * This component provides:
 * - Simple confirmation dialog with user name
 * - Keyboard navigation (Escape to cancel, Enter/Space to confirm)
 * - Auto-focus on cancel button for safety
 * - Accessibility features
 * - Clean, focused UI for important decisions
 * 
 * @component
 * @param {ConfirmDeleteModalProps} props - Component props
 * @returns {JSX.Element | null} Confirmation modal or null if not open
 * 
 * @example
 * ```tsx
 * import ConfirmDeleteModal from './components/ConfirmDeleteModal';
 * 
 * function UserList() {
 *   const [confirmOpen, setConfirmOpen] = useState(false);
 *   const [userToDelete, setUserToDelete] = useState<User | null>(null);
 * 
 *   const handleConfirmDelete = () => {
 *     // Delete user logic
 *     setConfirmOpen(false);
 *   };
 * 
 *   return (
 *     <ConfirmDeleteModal
 *       open={confirmOpen}
 *       onCancel={() => setConfirmOpen(false)}
 *       onConfirm={handleConfirmDelete}
 *       userName={userToDelete?.name || ''}
 *     />
 *   );
 * }
 * ```
 */
const ConfirmDeleteModal: React.FC<ConfirmDeleteModalProps> = ({ open, onCancel, onConfirm, userName }) => {
    /**
     * Sets up keyboard event listeners for accessibility.
     * 
     * This effect adds global keyboard listeners when the modal is open:
     * - Escape key: Cancels the deletion
     * - Enter or Space key: Confirms the deletion
     * 
     * The listeners are properly cleaned up when the modal closes or component unmounts.
     * 
     * @example
     * ```typescript
     * // Press Escape to cancel, Enter/Space to confirm
     * ```
     */
    React.useEffect(() => {
        if (!open) return;

        const handleKeyDown = (e: KeyboardEvent) => {
            if (e.key === 'Escape') onCancel();
            if (e.key === 'Enter' || e.key === ' ') onConfirm();
        };

        window.addEventListener('keydown', handleKeyDown);
        return () => window.removeEventListener('keydown', handleKeyDown);
    }, [open, onCancel, onConfirm]);

    // Don't render if modal is not open
    if (!open) return null;

    return (
        <div className="modal-backdrop confirm-backdrop" tabIndex={-1}>
            <div className="confirm-modal">
                <div className="confirm-title">Delete user?</div>
                <div className="confirm-message">Are you sure you want to delete <b>{userName}</b>?</div>
                <div className="confirm-actions">
                    <button className="confirm-btn cancel" onClick={onCancel} autoFocus>Cancel</button>
                    <button className="confirm-btn delete" onClick={onConfirm}>Delete</button>
                </div>
            </div>
        </div>
    );
};

export default ConfirmDeleteModal; 