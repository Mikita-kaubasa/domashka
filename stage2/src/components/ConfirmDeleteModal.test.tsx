import { fireEvent, render, screen } from '@testing-library/react';
import ConfirmDeleteModal from './ConfirmDeleteModal';

describe('ConfirmDeleteModal', () => {
    const userName = 'Alice';

    it('renders when open is true', () => {
        render(
            <ConfirmDeleteModal open={true} onCancel={jest.fn()} onConfirm={jest.fn()} userName={userName} />
        );
        expect(screen.getByText('Delete user?')).toBeInTheDocument();
        expect(
            screen.getByText((content, node) =>
                node?.textContent === `Are you sure you want to delete ${userName}?`
            )
        ).toBeInTheDocument();
    });

    it('does not render when open is false', () => {
        render(
            <ConfirmDeleteModal open={false} onCancel={jest.fn()} onConfirm={jest.fn()} userName={userName} />
        );
        expect(screen.queryByText('Delete user?')).not.toBeInTheDocument();
    });

    it('calls onCancel when Cancel button is clicked', () => {
        const handleCancel = jest.fn();
        render(
            <ConfirmDeleteModal open={true} onCancel={handleCancel} onConfirm={jest.fn()} userName={userName} />
        );
        fireEvent.click(screen.getByText('Cancel'));
        expect(handleCancel).toHaveBeenCalled();
    });

    it('calls onConfirm when Delete button is clicked', () => {
        const handleConfirm = jest.fn();
        render(
            <ConfirmDeleteModal open={true} onCancel={jest.fn()} onConfirm={handleConfirm} userName={userName} />
        );
        fireEvent.click(screen.getByText('Delete'));
        expect(handleConfirm).toHaveBeenCalled();
    });

    it('calls onCancel when Escape key is pressed', () => {
        const handleCancel = jest.fn();
        render(
            <ConfirmDeleteModal open={true} onCancel={handleCancel} onConfirm={jest.fn()} userName={userName} />
        );
        const modal = screen.getByText('Delete user?').closest('.modal-backdrop');
        if (modal) {
            fireEvent.keyDown(modal, { key: 'Escape', code: 'Escape' });
            expect(handleCancel).toHaveBeenCalled();
        }
    });

    it('calls onConfirm when Enter or Space key is pressed', () => {
        const handleConfirm = jest.fn();
        render(
            <ConfirmDeleteModal open={true} onCancel={jest.fn()} onConfirm={handleConfirm} userName={userName} />
        );
        const modal = screen.getByText('Delete user?').closest('.modal-backdrop');
        if (modal) {
            fireEvent.keyDown(modal, { key: 'Enter', code: 'Enter' });
            fireEvent.keyDown(modal, { key: ' ', code: 'Space' });
            expect(handleConfirm).toHaveBeenCalledTimes(2);
        }
    });

    it('has correct accessibility attributes', () => {
        render(
            <ConfirmDeleteModal open={true} onCancel={jest.fn()} onConfirm={jest.fn()} userName={userName} />
        );
        const modal = screen.getByText('Delete user?').closest('.modal-backdrop');
        expect(modal).toHaveAttribute('tabindex', '-1');
        const cancelBtn = screen.getByText('Cancel');
        const deleteBtn = screen.getByText('Delete');
        expect(document.activeElement).toBe(cancelBtn);
        expect(deleteBtn).toBeInTheDocument();
    });

    it('matches snapshot', () => {
        const { asFragment } = render(
            <ConfirmDeleteModal open={true} onCancel={jest.fn()} onConfirm={jest.fn()} userName={userName} />
        );
        expect(asFragment()).toMatchSnapshot();
    });
}); 