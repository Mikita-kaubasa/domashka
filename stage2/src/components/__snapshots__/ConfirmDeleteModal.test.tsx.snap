// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ConfirmDeleteModal matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="modal-backdrop confirm-backdrop"
    tabindex="-1"
  >
    <div
      class="confirm-modal"
    >
      <div
        class="confirm-title"
      >
        Delete user?
      </div>
      <div
        class="confirm-message"
      >
        Are you sure you want to delete 
        <b>
          Alice
        </b>
        ?
      </div>
      <div
        class="confirm-actions"
      >
        <button
          class="confirm-btn cancel"
        >
          Cancel
        </button>
        <button
          class="confirm-btn delete"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
